components/script_traits/lib.rs

```diff
@@ -1013,3 +1018,33 @@ impl ScriptToConstellationChan {
         self.sender.send((self.pipeline_id, msg))
     }
 }
+
+/// A data-holder for serialized data and transferred objects.
+/// <https://html.spec.whatwg.org/multipage/#structuredserializewithtransfer>
+#[derive(Debug, Deserialize, MallocSizeOf, Serialize)]
+pub struct StructuredSerializedData {
+    /// Data serialized by SpiderMonkey.
+    pub serialized: Vec<u8>,
+    /// Transferred objects.
+    pub ports: Option<HashMap<MessagePortId, MessagePortImpl>>,
```

Comment:
In the same vein as the options in `structuredclone`, while `serialized` will always be present, ports, and other transfereable in the future, will not. An empty hashmap could do the trick, and I felt that the option made it clearer when there are no transferred ports. 

Unlike the data holder in  `structuredclone`, it could be possible to make this an enum, as opposed to a struct with optional fields.